framework/src/hdgkernels/HDGKernel.C

```diff
@@ -0,0 +1,176 @@
+    _normals(_assembly.normals()),
+    _neigh(nullptr),
+    _current_side(_assembly.side()),
+    _computing_global_data(true),
+    _hibc_warehouse(
+        *getCheckedPointerParam<MooseObjectWarehouse<HDGIntegratedBC> *>("hibc_warehouse"))
+{
+  // This class handles residuals and Jacobians for multiple variables
+  _fe_problem.setKernelCoverageCheck(false);
+}
+
+void
+HDGKernel::addBCData(const HDGIntegratedBC & hibc)
+{
+  _PrimalMat += hibc._PrimalMat;
+  _LMMat += hibc._LMMat;
+  _PrimalLM += hibc._PrimalLM;
+  _LMPrimal += hibc._LMPrimal;
+  _PrimalVec += hibc._PrimalVec;
+  _LMVec += hibc._LMVec;
+}
+
+void
+HDGKernel::assemble()
+{
+  onElement();
+
+  for (const auto side : _current_elem->side_index_range())
+  {
+    // Whether we have reinitialized the finite element (libMesh FE) objects on the current face
+    bool fe_face_reinitd = false;
+
+    // Set up Sentinel class so that, even if reinitMaterialsFace() throws, we
+    // still remember to swap back during stack unwinding. Stack is LIFO so this will be destructed
+    // before fe_face_reinitd, so it's valid for us to pass it in here
+    SwapBackSentinel sentinel(
+        _fe_problem, &FEProblemBase::swapBackMaterialsFace, _tid, fe_face_reinitd, false);
+
+    const auto & boundary_ids = _mesh.getBoundaryIDs(_current_elem, side);
+    for (const auto bnd_id : boundary_ids)
+      if (_hibc_warehouse.hasActiveBoundaryObjects(bnd_id, _tid))
+      {
+        _fe_problem.setCurrentBoundaryID(bnd_id, _tid);
+
+        if (!fe_face_reinitd)
+        {
+          fe_face_reinitd = true;
+          // Do the whole shebang:
+          // - reinit FE objects on the face
+          // - compute variable values on the face
+          // - compute material values, including values from face and boundary materials
+          NonlinearThread::prepareFace(_fe_problem, _tid, _current_elem, side, _current_bnd_id);
+        }
+        else
+          // Only need to compute boundary materials; face materials have already been computed on
+          // the current face
+          _fe_problem.reinitMaterialsBoundary(_current_bnd_id, _tid);
+
+        libmesh_assert(_current_side == side);
+        const auto & bcs = _hibc_warehouse.getActiveBoundaryObjects(bnd_id, _tid);
+        for (const auto & bc : bcs)
+        {
+          mooseAssert(bc->shouldApply(),
+                      "I don't think anyone uses the shouldApply feature for integrated boundary "
+                      "conditions");
+          bc->onBoundary();
+          addBCData(*bc);
+        }
+      }
+
+    if (_neigh = _current_elem->neighbor_ptr(side);
+        _neigh && this->hasBlocks(_neigh->subdomain_id()))
+    {
+      if (!fe_face_reinitd)
+      {
+        fe_face_reinitd = true;
+        NonlinearThread::prepareFace(
+            _fe_problem, _tid, _current_elem, side, Moose::INVALID_BOUNDARY_ID);
+      }
+
+      libmesh_assert(_current_side == side);
+      onInternalSide();
+    }
+  }
+
```

Comment:
there's two parts to this assemble routine

let s split them with a comment like:
// Now that the HDGData members are filled, we contribute them to ..